chips/stm32f303xc/src/adc.rs

```diff
@@ -0,0 +1,740 @@
+    ThreeChannels = 0b010,
+    FourChannels = 0b011,
+    FiveChannels = 0b100,
+    SixChannels = 0b101,
+    SevenChannels = 0b110,
+    EightChannels = 0b111,
+}
+
+#[allow(dead_code)]
+#[repr(u32)]
+enum ExternalTriggerDetection {
+    Disabled = 0b00,
+    RisingEdge = 0b01,
+    FallingEdge = 0b10,
+    RisingAndFalling = 0b11,
+}
+
+#[allow(dead_code)]
+#[repr(u32)]
+enum ExternalTriggerSelection {
+    Event0 = 0b0000,
+    Event1 = 0b0001,
+    Event2 = 0b0010,
+    Event3 = 0b0011,
+    Event4 = 0b0100,
+    Event5 = 0b0101,
+    Event6 = 0b0110,
+    Event7 = 0b0111,
+    Event8 = 0b1000,
+    Event9 = 0b1001,
+    Event10 = 0b1010,
+    Event11 = 0b1011,
+    Event12 = 0b1100,
+    Event13 = 0b1101,
+    Event14 = 0b1110,
+    Event15 = 0b1111,
+}
+
+#[allow(dead_code)]
+#[repr(u32)]
+enum DataResolution {
+    Bit12 = 0b00,
+    Bit10 = 0b01,
+    Bit8 = 0b10,
+    Bit6 = 0b11,
+}
+
+#[derive(Copy, Clone, PartialEq)]
+enum ADCStatus {
+    Idle,
+    Off,
+    PoweringOn,
+    OneSample,
+}
+
+pub struct Adc {
+    registers: StaticRef<AdcRegisters>,
+    common_registers: StaticRef<AdcCommonRegisters>,
+    clock: AdcClock,
+    status: Cell<ADCStatus>,
+    client: OptionalCell<&'static dyn EverythingClient>,
+    // clock34: AdcClock,
+}
+
+pub static mut ADC1: Adc = Adc::new();
+
+impl Adc {
+    const fn new() -> Adc {
+        Adc {
+            registers: ADC1_BASE,
+            //rcc copy
+            common_registers: ADC12_COMMON_BASE,
+            clock: AdcClock(rcc::PeripheralClock::AHB(rcc::HCLK::ADC1)),
+            // clock34: AdcClock(rcc::PeripheralClock::AHB(rcc::HCLK::ADC34)),
+            status: Cell::new(ADCStatus::Off),
+            client: OptionalCell::empty(),
+        }
+    }
+
+    pub fn enable(&self) {
+        self.status.set(ADCStatus::PoweringOn);
+
+        // Enable adc clock
+        self.enable_clock();
+
+        //Set Synchronous clock mode
+        self.common_registers.ccr.modify(CCR::CKMODE.val(0b01));
+
+        self.registers.cr.modify(CR::ADVREGEN.val(0b00));
+        self.registers.cr.modify(CR::ADVREGEN.val(0b01));
+
+        // Wait for ADVRGEN to enable
+        // This needs to be synchronous because there is no interrupt signaling
+        // when ADVRGEN becomes enabled
+        for _i in 0..1000000 {
+            unsafe {
+                llvm_asm!(
+                "nop"
+            : : : : "volatile" );
+            }
+        }
+
+        // Enable the temperature sensor
+        self.common_registers.ccr.modify(CCR::TSEN::SET);
+
+        // Enable ADC Ready interrupt
+        self.registers.ier.modify(IER::ADRDYIE::SET);
+
+        // Clear registers
+        self.registers.isr.modify(ISR::ADRDY::CLEAR);
+        self.registers.cr.modify(CR::ADEN::CLEAR);
+        self.registers.cr.modify(CR::ADCALDIF::CLEAR);
+        self.registers.cr.modify(CR::ADCAL::SET);
+
+        // Wait for calibration
+        while self.registers.cr.is_set(CR::ADCAL) {}
```

Comment:
Do you have a sense of how many cycles this wait should take? As a general rule, we favor asynchronous designs unless the wait is expected to be very small.

Reply:
No, there is not a fixed amount of cycles needed that's why I chose busy waiting for this to wait the completion

Reply:
This is the link to the reference manual for stm32f3xx https://www.st.com/resource/en/reference_manual/dm00043574-stm32f303xb-c-d-e-stm32f303x6-8-stm32f328x8-stm32f358xc-stm32f398xe-advanced-arm-based-mcus-stmicroelectronics.pdf and on page 317 it is described how to do the calibration process. It says that the ADCAL bit will be cleared by the hardware as soon as the process is completed.